cocotb/share/makefiles/simulators/Makefile.verilator

```diff
@@ -0,0 +1,96 @@
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of Potential Ventures Ltd,
+#       SolarFlare Communications Inc nor the
+#       names of its contributors may be used to endorse or promote products
+#       derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL POTENTIAL VENTURES LTD BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+
+ifneq ($(VHDL_SOURCES),)
+
+results.xml:
+	@echo "Skipping simulation as VHDL is not supported on simulator=$(SIM)"
+debug: results.xml
+clean::
+
+else
+
+CMD := verilator
+
+ifeq ($(shell which $(CMD) 2>/dev/null),)
+# Verilator is not in PATH, lets start searching for it
+  ifdef VERILATOR_ROOT
+    CMD = $(VERILATOR_ROOT)/bin/verilator
+  else
+    PKG_CONFIG ?= pkg-config
+    ifneq ($(shell which $(PKG_CONFIG) 2>/dev/null),)
+      # Use pkg-config to find it
+      EXEC = $(shell $(PKG_CONFIG) --variable=exec_prefix verilator)
+      ifeq ($(EXEC),)
+        $(error Cannot find verilator, either set VERILATOR_ROOT or add it to PKG_CONFIG_PATH.)
+      endif
+      CMD = $(EXEC)/bin/verilator
+    else
+        $(error Cannot find verilator, either set VERILATOR_ROOT or add it to PKG_CONFIG_PATH.)
+    endif
+  endif
+endif
+
+COMPILE_ARGS += --vpi -LDFLAGS "-L $(PWD)/build/libs/x86_64/ -lvpi -lgpi -lcocotb -lgpilog -lcocotbutils"
+
+$(SIM_BUILD)/verilator.cpp: $(COCOTB_SHARE_DIR)/lib/verilator/verilator.cpp.in $(SIM_BUILD)
+	sed -e 's/%TOPLEVEL%/$(TOPLEVEL)/g' $< > $@
+
+# Compilation phase
+$(SIM_BUILD)/V$(TOPLEVEL): $(SIM_BUILD)/verilator.cpp $(COCOTB_LIBS) $(COCOTB_VPI_LIB) $(VERILOG_SOURCES) $(CUSTOM_COMPILE_DEPS)
+	$(CMD) -cc --exe -Mdir $(SIM_BUILD) -DCOCOTB_SIM=1 --top-module $(TOPLEVEL) $(COMPILE_ARGS) $(EXTRA_ARGS) $(VERILOG_SOURCES) $(SIM_BUILD)/verilator.cpp
+	make -C $(SIM_BUILD) -f V$(TOPLEVEL).mk
```

Comment:
It would be slightly cleaner and potentially reduce recompilation steps if this target is split into two: one for the verilator call, and the other one for the make invocation.

Additionally, can you add a variable for arguments passed to make? I often need to pass `OPT_FAST="-O2"` to verilator, and the only way to do that is through the Makefile.

Reply:
I will add the second one later.

Reply:
Everything else is resolved. Further improvements and functionalities added. Verilator will land the next couple of days, we are done then for the initial release!